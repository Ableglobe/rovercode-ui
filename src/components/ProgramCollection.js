import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import { hot } from 'react-hot-loader';
import {
  Button,
  Card,
  Grid,
  Header,
  Input,
} from 'semantic-ui-react';

import CustomPagination from './CustomPagination';

class ProgramCollection extends PureComponent {
  handlePageChange = (e, { activePage }) => {
    const { onPageChange, owned } = this.props;

    onPageChange(activePage, owned);
  }

  handleSearchChange = (event) => {
    const { onSearchChange, owned } = this.props;

    onSearchChange(event.target.value, owned);
  }

  render() {
    const {
      label,
      onProgramClick,
      onRemoveClick,
      programs,
      owned,
    } = this.props;

    return (
      <Fragment>
        <Grid centered columns={3}>
          <Grid.Row>
            <Grid.Column />
            <Grid.Column>
              <Header as="h1" textAlign="center">
                {label}
              </Header>
            </Grid.Column>
            <Grid.Column textAlign="right">
              <Input
                className="prompt"
                icon="search"
                placeholder="Search..."
                onChange={this.handleSearchChange}
              />
            </Grid.Column>
          </Grid.Row>
        </Grid>
        <Card.Group centered>
          {
            programs.results.map(program => (
              <Card key={program.id}>
                <Card.Content>
                  <Card.Header>
                    {program.name}
                  </Card.Header>
                  <Card.Meta>
                    { owned ? 'Mine' : program.user.username }
                  </Card.Meta>
                </Card.Content>
                <Card.Content extra>
                  <Button primary id={program.id} onClick={onProgramClick}>
                    { owned ? 'Keep Working' : 'View' }
                  </Button>
                  {
                    owned ? (
                      <Button
                        negative
                        id={program.id}
                        name={program.name}
                        onClick={onRemoveClick}
                        floated="right"
                      >
                        Remove
                      </Button>
                    ) : (null)
                  }
                </Card.Content>
              </Card>
            ))
          }
        </Card.Group>
        {
          programs.total_pages > 1 ? (
            <Grid centered>
              <Grid.Row>
                <CustomPagination
                  defaultActivePage={1}
                  totalPages={programs.total_pages}
                  onPageChange={this.handlePageChange}
                />
              </Grid.Row>
            </Grid>
          ) : (null)
        }
      </Fragment>
    );
  }
}

ProgramCollection.defaultProps = {
  owned: false,
};

ProgramCollection.propTypes = {
  programs: PropTypes.shape({
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        user: PropTypes.shape({
          username: PropTypes.string.isRequired,
        }).isRequired,
      }),
    ),
  }).isRequired,
  label: PropTypes.string.isRequired,
  owned: PropTypes.bool,
  onProgramClick: PropTypes.func.isRequired,
  onRemoveClick: PropTypes.func.isRequired,
  onPageChange: PropTypes.func.isRequired,
  onSearchChange: PropTypes.func.isRequired,
};

export default hot(module)(ProgramCollection);
