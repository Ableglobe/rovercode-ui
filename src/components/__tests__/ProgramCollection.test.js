import React from 'react';
import {
  Button,
  Card,
  Header,
  Input,
} from 'semantic-ui-react';
import { shallow } from 'enzyme';
import CustomPagination from '../CustomPagination';
import ProgramCollection from '../ProgramCollection';

let onProgramClick;
let onRemoveClick;
let onPageChange;
let onSearchChange;

describe('The ProgramCollection component', () => {
  beforeEach(() => {
    onProgramClick = jest.fn();
    onRemoveClick = jest.fn();
    onPageChange = jest.fn();
    onSearchChange = jest.fn();
  });

  test('renders on the page with no errors', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 2,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: 10,
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: 1,
      }],
    };
    const wrapper = shallow(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        owned
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onPageChange={onPageChange}
        onSearchChange={onSearchChange}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('shows the correct number of programs for the user', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: 1,
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: 1,
      }],
    };
    const wrapper = shallow(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        owned
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onPageChange={onPageChange}
        onSearchChange={onSearchChange}
      />,
    );

    expect(wrapper.find(Header).exists()).toBe(true);
    expect(wrapper.find(Card).length).toBe(2);
    expect(wrapper.find(Card.Meta).first().prop('children')).toBe('Mine');
    expect(wrapper.find(Button).first().prop('children')).toBe('Keep Working');
  });

  test('shows the correct number of programs for other users', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: 1,
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: 1,
      }],
    };
    const wrapper = shallow(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onPageChange={onPageChange}
        onSearchChange={onSearchChange}
      />,
    );

    expect(wrapper.find(Header).exists()).toBe(true);
    expect(wrapper.find(Card).length).toBe(2);
    expect(wrapper.find(Card.Meta).first().prop('children')).toBe(1);
    expect(wrapper.find(Button).first().prop('children')).toBe('View');
  });

  test('callback when program click', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: 10,
      }],
    };
    const wrapper = shallow(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        owned
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onPageChange={onPageChange}
        onSearchChange={onSearchChange}
      />,
    );

    wrapper.find(Button).first().simulate('click', {
      target: {
        id: 33,
      },
    });

    expect(onProgramClick).toHaveBeenCalledWith({
      target: {
        id: 33,
      },
    });
  });

  test('callback when remove click', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: 10,
      }],
    };
    const wrapper = shallow(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        owned
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onPageChange={onPageChange}
        onSearchChange={onSearchChange}
      />,
    );

    wrapper.find(Button).last().simulate('click', {
      target: {
        id: 33,
      },
    });

    expect(onRemoveClick).toHaveBeenCalledWith({
      target: {
        id: 33,
      },
    });
  });

  test('callback when page changes', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 2,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: 1,
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: 1,
      }],
    };
    const wrapper = shallow(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        owned
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onPageChange={onPageChange}
        onSearchChange={onSearchChange}
      />,
    );

    wrapper.find(CustomPagination).simulate('pageChange', null, {
      activePage: 2,
    });

    expect(onPageChange).toHaveBeenCalledWith(2, true);
  });

  test('callback when search changes', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 2,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: 1,
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: 1,
      }],
    };
    const wrapper = shallow(
      <ProgramCollection
        programs={programs}
        label="My Programs"
        owned
        onProgramClick={onProgramClick}
        onRemoveClick={onRemoveClick}
        onPageChange={onPageChange}
        onSearchChange={onSearchChange}
      />,
    );

    wrapper.find(Input).simulate('change', {
      target: {
        value: 'abc',
      },
    });

    expect(onSearchChange).toHaveBeenCalledWith('abc', true);
  });
});
