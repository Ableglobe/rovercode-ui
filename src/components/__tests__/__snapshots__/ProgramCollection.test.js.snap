// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ProgramCollection component renders on the page with no errors 1`] = `
<Fragment>
  <Grid
    centered={true}
    columns={5}
  >
    <GridRow>
      <GridColumn />
      <GridColumn />
      <GridColumn>
        <Header
          as="h1"
          textAlign="center"
        >
          My Programs
        </Header>
      </GridColumn>
      <GridColumn
        textAlign="right"
      >
        <Input
          className="prompt"
          icon="search"
          onChange={[Function]}
          placeholder="Search..."
          type="text"
        />
      </GridColumn>
      <GridColumn>
        <Dropdown
          additionLabel="Add "
          additionPosition="top"
          button={true}
          className="icon"
          closeOnBlur={true}
          deburr={false}
          floating={true}
          icon="sort"
          labeled={true}
          minCharacters={1}
          noResultsMessage="No results found."
          openOnFocus={true}
          renderLabel={[Function]}
          searchInput="text"
          selectOnBlur={true}
          selectOnNavigation={true}
          text="Sort"
          wrapSelection={true}
        >
          <DropdownMenu>
            <DropdownItem
              name="name"
              onClick={[Function]}
            >
              Name
              <Icon
                as="i"
                name="caret down"
              />
            </DropdownItem>
          </DropdownMenu>
        </Dropdown>
      </GridColumn>
    </GridRow>
  </Grid>
  <CardGroup
    centered={true}
  >
    <Card
      key="33"
    >
      <CardContent>
        <CardHeader>
          Unnamed_Design_3
        </CardHeader>
        <CardMeta>
          Mine
        </CardMeta>
      </CardContent>
      <CardContent
        extra={true}
      >
        <Button
          as="button"
          data-owned={true}
          id={33}
          onClick={[MockFunction]}
          primary={true}
        >
          Keep Working
        </Button>
        <Button
          as="button"
          floated="right"
          id={33}
          name="Unnamed_Design_3"
          negative={true}
          onClick={[MockFunction]}
        >
          Remove
        </Button>
      </CardContent>
    </Card>
    <Card
      key="5"
    >
      <CardContent>
        <CardHeader>
          Unnamed_Design_2
        </CardHeader>
        <CardMeta>
          Mine
        </CardMeta>
      </CardContent>
      <CardContent
        extra={true}
      >
        <Button
          as="button"
          data-owned={true}
          id={5}
          onClick={[MockFunction]}
          primary={true}
        >
          Keep Working
        </Button>
        <Button
          as="button"
          floated="right"
          id={5}
          name="Unnamed_Design_2"
          negative={true}
          onClick={[MockFunction]}
        >
          Remove
        </Button>
      </CardContent>
    </Card>
  </CardGroup>
  <Grid
    centered={true}
  >
    <GridRow>
      <CustomPagination
        defaultActivePage={1}
        onPageChange={[Function]}
        totalPages={2}
      />
    </GridRow>
  </Grid>
</Fragment>
`;
